zerver/lib/message.py

```diff
@@ -577,9 +612,23 @@ def bulk_access_stream_messages_query(
     if stream.is_public() and user_profile.can_access_public_streams():
         return messages
 
-    if not Subscription.objects.filter(
-        user_profile=user_profile, active=True, recipient=stream.recipient
-    ).exists():
+    def has_content_access_helper() -> bool:
+        is_subscribed = Subscription.objects.filter(
+            user_profile=user_profile, active=True, recipient=stream.recipient
+        ).exists()
+
+        if is_subscribed:
+            return True
+
+        user_recursive_group_ids = set(
+            get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+        )
+
+        return is_user_in_groups_granting_content_access(stream, user_recursive_group_ids)
```

Comment:
Is this the same function you have above? It could make sense to make it a top-level shared function called from both places.